src/access/AccessRestrictionProvider.h

```diff
@@ -0,0 +1,225 @@
+    virtual ~AccessRestrictionProvider() = default;
+
+    AccessRestrictionProvider(const AccessRestrictionProvider &)             = delete;
+    AccessRestrictionProvider & operator=(const AccessRestrictionProvider &) = delete;
+
+    /**
+     * Set the restriction entries that are to be used during commissioning when there is no accessing fabric.
+     *
+     * @param [in] entries  The entries to set.
+     */
+    CHIP_ERROR SetCommissioningEntries(const std::vector<Entry> & entries);
+
+    /**
+     * Set the restriction entries for a fabric.
+     *
+     * @param [in] fabricIndex  The index of the fabric for which to create entries.
+     * @param [in] entries  The entries to set for the fabric.
+     */
+    CHIP_ERROR SetEntries(const FabricIndex, const std::vector<Entry> & entries);
+
+    /**
+     * Add a listener to be notified of changes in the access restriction list and active reviews.
+     *
+     * @param [in] listener  The listener to add.
+     */
+    void AddListener(Listener & listener);
+
+    /**
+     * Remove a listener from being notified of changes in the access restriction list and active reviews.
+     *
+     * @param [in] listener  The listener to remove.
+     */
+    void RemoveListener(Listener & listener);
+
+    /**
+     * Check whether access by a subject descriptor to a request path should be restricted (denied) for the given action.
+     * These restrictions are are only a part of overall access evaluation.
+     *
+     * If access is not restricted, CHIP_NO_ERROR will be returned.
+     *
+     * @retval CHIP_ERROR_ACCESS_DENIED if access is denied.
+     * @retval other errors should also be treated as restricted/denied.
+     * @retval CHIP_NO_ERROR if access is not restricted/denied.
+     */
+    CHIP_ERROR Check(const SubjectDescriptor & subjectDescriptor, const RequestPath & requestPath);
+
+    /**
+     * Request a review of the access restrictions for a fabric.
+     *
+     * @param [in]  fabricIndex  The index of the fabric requesting a review.
+     * @param [in]  arl          An optinal list of access restriction entries to review.  If null, all entries will be reviewed.
+     * @param [out] token        The unique token for the review, which can be matched to a review update event.
+     */
+    CHIP_ERROR RequestFabricRestrictionReview(FabricIndex fabricIndex, const std::vector<Entry> & arl, uint64_t & token)
+    {
+        token = ++mNextToken;
+        return DoRequestFabricRestrictionReview(fabricIndex, token, arl);
+    }
+
+    /**
+     * Get the commissioning restriction entries.
+     *
+     * @retval the commissioning restriction entries.
+     */
+    const std::vector<Entry> & GetCommissioningEntries() const { return mCommissioningEntries; }
+
+    /**
+     * Get the restriction entries for a fabric.
+     *
+     * @param [in]  fabricIndex the index of the fabric for which to get entries.
+     * @param [out] entries reference to a const vector to hold the entries.
```

Comment:
Needs to clearly say how long the pointer is valid for.

Reply:
Fixed